storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java

```diff
@@ -158,35 +214,37 @@ public BlobClientBuilder endpoint(String endpoint) {
             if (parts.snapshot() != null) {
                 this.snapshot = parts.snapshot();
             }
```

Comment:
So I think this branch condition is a little off (once this method also uses the path of the URL). Take this set of steps:

1. `myClient1 = builder.endpoint("http://myaccount.blob.core.windows.net/mycontainer/myblob?snapshot=mysnapshotid").buildBlobClient();`
2. `myClient2 = builder.endpoint("http://myaccount.blob.core.windows.net/anothercontainer/anotherblob").buildBlobClient();`

Step 2 will NOT erase the snapshot in the builder, and myClient2 will point to `http://myaccount.blob.core.windows.net/anothercontainer/anotherblob?snapshot=mysnapshotid`, at best a nonexistent resource and at worst the wrong one.

We need to make decisions about what we overwrite under what circumstances when this method is used, as well as clearly document it. My first thoughts include:

1. Host and scheme are always overwritten
2. If any part of the path is specified, then we overwrite container name, blob name, and snapshot id. (if there is only one path segment, we assume it's the implicit root container) (we don't need to worry about only the snapshot being specified, as you can't snapshot an account)
3. SAS token is always overwritten, including setting it to null if not present

Input from @rickle-msft and @JonathanGiles is much appreciated.

Reply:
I agree with those proposals, they are not only easier to explain but also removes assumptions we are trying to make on the behalf of the consumer.